### ISSUE TITLE
Integrate `churn` Gem into Rubycritic to Speed Up RubyCritic Churn Computation

### ISSUE THREAD

**Issue by hemalvarambhia** [whitesmith/rubycritic#498]
Problem: I use `rubycritic` on some repositories and some take 15+ minutes, which seems similar to #490.

Solution: ~I~ @Greg-Myers-SB found this gem by Dan Mayer (see [Churn's Github page](https://github.com/danmayer/churn/blob/main/README.md)) and found it was a lot faster. Can using the said gem happen?

**Comment by Greg-Myers-SB** [comment]
On the same codebase.

Rubycritic's Churn:
![image](https://github.com/user-attachments/assets/5035a2f9-df05-49a0-ae66-9f2a391a5687)
<img width="150" alt="image" src="https://github.com/user-attachments/assets/5c9c4303-d990-42e3-b44e-74d3321f17c0">

Churn gem:
![image](https://github.com/user-attachments/assets/d9a5987d-2d0e-4f56-971e-1ec842b99cb1)
![image](https://github.com/user-attachments/assets/e4cfdd90-8446-46f9-8527-e65e707cfa4c)

Churn gem is _significantly_ faster, we might be able to make an adapter for other Churn sources and figure out what the major differences in results are.

**Comment by Greg-Myers-SB** [comment]
Yep, of course it's not just speed, the churn gem is not a 1-1 with the current implementation and we may spend some time into looking at what the differences are.

**Comment by hemalvarambhia** [comment]
@Greg-Myers-SB I saw that it has a hg analyser module. I thought Hg and mercurial are the same. How wrong am I?

**Comment by hemalvarambhia** [comment]
> Ideally we want to keep the rubycritic test suite as is when switching to churn -- I'm not totally sure that there are good enough tests for the churn behavior though.

@etagwerker and what kind of good enough is that? Is this something you would like addressed?

Edit: I'd like to credit @Greg-Myers-SB for finding the gem.

**Comment by etagwerker** [comment]
Also, just FYI, I don't think we need a release of `churn` to see this working: https://github.com/danmayer/churn/pull/88#issuecomment-2511088428

I envision the integration will not use `churn`'s JSON output, but it will integrate programmatically with `churn`

**Comment by hemalvarambhia** [comment]
I have a small PR, in which I add a learning test to understand how the gem works. In case anyone's not familiar with the idea of a learning test, it is to way to gain clarification on how an external library and has the bonus of documenting assumptions that can be tested when the gem is updated.

**Comment by hemalvarambhia** [comment]
@etagwerker #499 contains a learning test (Test-Driven Development By Example, Kent Beck) I thought I could add as a safe microstep towards contributing to Rubycritic. It is small enough to throw away if I need to start over and hopefully easy for you to review and safe enough consider making part of the codebase.

**Comment by mateusdeap** [comment]
@Greg-Myers-SB good to hear.

However, all the work we did around understanding this module kind of got me thinking that maybe we should rethink this metric.

One of the things that stood out to me was: we're calculating churn starting at the beginning of a file's commit history, no matter how long ago that was.

The thing is this metric is only good if we presume that a file's changes are evenly distributed in time. This means that a file that is, say, 8 years old, and saw _a lot_ of change during it's first 4 years, can score a very high churn and yet be a non issue or definitely not fall into the category of "file that is changed often" (which is what I presume to be the point of this metric).

This impression was further reinforced by the fact that the churn gem puts a limit on how far back it goes by default. I'm thinking that might actually be a useful thing to have here as well.

Furthermore I'd like to even propose a slightly different metric: churn rate or average churn rate, which measure how much churn a file has over a period of time. The question would be: what would be a sensible period of time?

I feel like it's a separate discussion and I can open an RFC here in the repo if there's interest. What do you think?

cc.: @etagwerker, @hmdros, @julioalucero

**Comment by etagwerker** [comment]
Closing this for now. If an integration with the `churn` gem would make things faster, I'd be open to re-opening this issue.
